File: src/run_qwen3_batch.ts
import { GlmOps } from "./glm_ops.js";
import { Qwen3Model } from "./qwen3_model.js";
import { PagedKVCache, WorkspaceBuffers } from "./paged_kv.js";
import { AutoTokenizer } from "@huggingface/transformers";
import { resolveModelPath } from "./model_path.js";
import { createInterface } from "node:readline";

const QWEN3_REPO = "Qwen/Qwen3-0.6B";
const EOS_TOKEN_IDS = new Set([151645, 151643]);

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  let maxTokens = 128;
  let noThink = false;
  let gpu: number | undefined;
  let maxSeqLen = 2048;
  let maxPages = 128;
  let maxBatch = 4;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--max-tokens" && i + 1 < args.length) {
      maxTokens = parseInt(args[++i], 10);
    } else if (args[i] === "--no-think") {
      noThink = true;
    } else if (args[i] === "--gpu" && i + 1 < args.length) {
      gpu = parseInt(args[++i], 10);
    } else if (args[i] === "--max-seq-len" && i + 1 < args.length) {
      maxSeqLen = parseInt(args[++i], 10);
    } else if (args[i] === "--max-pages" && i + 1 < args.length) {
      maxPages = parseInt(args[++i], 10);
    } else if (args[i] === "--max-batch" && i + 1 < args.length) {
      maxBatch = parseInt(args[++i], 10);
    }
  }

  const gpuId = gpu ?? parseInt(process.env.GLM_GPU ?? "0", 10);
  process.env.CUDA_VISIBLE_DEVICES = String(gpuId);

  console.log(`Loading model on GPU ${gpuId}...`);
  const glm = new GlmOps(0);
  const model = Qwen3Model.fromPretrained(glm, QWEN3_REPO, maxBatch, maxSeqLen);

  const modelDir = resolveModelPath(QWEN3_REPO);
  const tokenizer = await AutoTokenizer.from_pretrained(modelDir, { local_files_only: true });

  const cfg = (model as any).cfg;
  const nKv = cfg.numKeyValueHeads;
  const hd = cfg.headDim;
  const nLayers = cfg.numHiddenLayers;

  const ws = new WorkspaceBuffers(glm);
  const pagedKV = new PagedKVCache(glm, nKv, hd, nLayers, maxPages, maxBatch);

  const enableThinking = !noThink;

  console.log(`Qwen3-0.6B batched ready (max_tokens=${maxTokens}, thinking=${noThink ? "off" : "on"})`);
  console.log("Enter prompts one per line. Empty line to submit batch. /clear to reset, /q to quit.");

  const rl = createInterface({ input: process.stdin, output: process.stdout });
  const askLine = (prompt: string): Promise<string> =>
    new Promise((resolve) => rl.question(prompt, resolve));

  try {
    while (true) {
      const prompts: string[] = [];
      while (true) {
        const line = (await askLine(`\nPrompt ${prompts.length + 1} (empty=done): `)).trim();
        if (line === "") {
          break;
        }
        if (line.toLowerCase() === "quit" || line.toLowerCase() === "exit" || line.toLowerCase() === "/q") {
          rl.close();
          return;
        }
        if (line.toLowerCase() === "/clear") {
          prompts.length = 0;
          console.log("Batch cleared.");
          continue;
        }
        prompts.push(line);
      }

      if (prompts.length === 0) continue;

      const inputIdsList: number[][] = [];
      for (const prompt of prompts) {
        const messages = [{ role: "user" as const, content: prompt }];
        const result = tokenizer.apply_chat_template(messages as any, {
          tokenize: true,
          add_generation_prompt: true,
          return_tensor: false,
          return_dict: true,
          tokenizer_kwargs: { enable_thinking: enableThinking },
        }) as { input_ids: number[] | number[][] };
        const ids = (Array.isArray(result.input_ids[0]) ? result.input_ids[0] : result.input_ids) as number[];
        inputIdsList.push(ids);
      }

      const maxPromptLen = Math.max(...inputIdsList.map(ids => ids.length));
      if (maxPromptLen > maxSeqLen) {
        console.log(`Longest prompt (${maxPromptLen} tokens) exceeds max_seq_len (${maxSeqLen}). Skipping.`);
        continue;
      }

      const start = Date.now();
      const generatedIds = model.generateBatch(inputIdsList, ws, pagedKV, maxTokens, EOS_TOKEN_IDS);
      const elapsed = (Date.now() - start) / 1000;
      const totalTokens = generatedIds.reduce((sum, ids) => sum + ids.length, 0);

      for (let i = 0; i < prompts.length; i++) {
        const text = tokenizer.decode(generatedIds[i], { skip_special_tokens: true });
        console.log(`\n--- Response ${i + 1} ---`);
        console.log(text);
      }

      console.log(`\n  [${prompts.length} prompts, ${totalTokens} tokens, ${elapsed.toFixed(1)}s, ${(totalTokens / elapsed).toFixed(1)} tok/s]`);
    }
  } finally {
    pagedKV.free();
    ws.free();
    model.free();
    rl.close();
  }
}

main().catch((err) => {
  console.error("Failed:", err);
  process.exit(1);
});

File: tests/python/run_qwen3_batch.py
#!/usr/bin/env python3
import argparse
import os
import sys
import time

sys.path.insert(0, os.path.dirname(__file__))
from helpers import GlmOps
from qwen3_model import Qwen3Model, _normalize_token_ids
from paged_kv import PagedKVCache, WorkspaceBuffers

from transformers import AutoTokenizer

QWEN3_REPO = "Qwen/Qwen3-0.6B"
EOS_TOKEN_IDS = {151645, 151643}


def main():
    parser = argparse.ArgumentParser(description="Batched interactive Qwen3-0.6B chat")
    parser.add_argument("--max-tokens", type=int, default=128, help="Max tokens per response")
    parser.add_argument("--no-think", action="store_true", help="Disable thinking mode")
    parser.add_argument("--gpu", type=int, default=None, help="GPU device ID")
    parser.add_argument("--max-seq-len", type=int, default=2048, help="Max sequence length")
    parser.add_argument("--max-pages", type=int, default=128, help="Max pages for paged KV cache")
    parser.add_argument("--max-batch", type=int, default=4, help="Max batch size")
    args = parser.parse_args()

    gpu_id = args.gpu if args.gpu is not None else int(os.environ.get("GLM_GPU", "0"))
    os.environ["CUDA_VISIBLE_DEVICES"] = str(gpu_id)

    print(f"Loading model on GPU {gpu_id}...")
    glm = GlmOps(device_id=0)
    model = Qwen3Model.from_pretrained(glm, QWEN3_REPO, max_batch=args.max_batch, max_seq_len=args.max_seq_len)
    tokenizer = AutoTokenizer.from_pretrained(QWEN3_REPO)

    cfg = model.cfg
    n_kv = cfg.num_key_value_heads
    hd = cfg.head_dim
    n_layers = cfg.num_hidden_layers

    ws = WorkspaceBuffers(glm)
    paged_kv = PagedKVCache(glm, n_kv, hd, n_layers, args.max_pages, max_batch=args.max_batch)

    enable_thinking = not args.no_think

    print(f"Qwen3-0.6B batched ready (max_tokens={args.max_tokens}, thinking={'off' if args.no_think else 'on'})")
    print("Enter prompts separated by blank lines. Double blank line to submit batch.")
    print("/clear to reset, /q to quit.")

    try:
        while True:
            prompts = []
            try:
                while True:
                    line = input(f"\nPrompt {len(prompts) + 1} (blank=done, double blank=submit): ")
                    if line.strip() == "":
                        if prompts:
                            break
                        continue
                    if line.strip().lower() in {"quit", "exit", "/q"}:
                        raise SystemExit(0)
                    if line.strip().lower() == "/clear":
                        prompts = []
                        print("Batch cleared.")
                        continue
                    prompts.append(line.strip())
            except (EOFError, KeyboardInterrupt):
                print()
                break

            if not prompts:
                continue

            input_ids_list = []
            for prompt in prompts:
                messages = [{"role": "user", "content": prompt}]
                try:
                    ids = tokenizer.apply_chat_template(
                        messages, tokenize=True, add_generation_prompt=True,
                        enable_thinking=enable_thinking,
                    )
                except TypeError:
                    ids = tokenizer.apply_chat_template(
                        messages, tokenize=True, add_generation_prompt=True,
                    )
                ids = _normalize_token_ids(ids, tokenizer)
                input_ids_list.append(ids)

            max_prompt_len = max(len(ids) for ids in input_ids_list)
            if max_prompt_len > args.max_seq_len:
                print(f"Longest prompt ({max_prompt_len} tokens) exceeds max_seq_len ({args.max_seq_len}). Skipping.")
                continue

            start = time.time()
            generated_ids = model.generate_batch(
                input_ids_list, ws, paged_kv,
                max_new_tokens=args.max_tokens,
                eos_token_ids=EOS_TOKEN_IDS,
            )
            elapsed = time.time() - start
            total_tokens = sum(len(ids) for ids in generated_ids)

            for i, (prompt, ids) in enumerate(zip(prompts, generated_ids)):
                text = tokenizer.decode(ids, skip_special_tokens=True)
                print(f"\n--- Response {i + 1} ---")
                print(text)

            print(f"\n  [{len(prompts)} prompts, {total_tokens} tokens, {elapsed:.1f}s, {total_tokens/elapsed:.1f} tok/s]")
    finally:
        paged_kv.free()
        ws.free()
        model.free()


if __name__ == "__main__":
    main()
